File: admin/services/socket.service.ts
import { io, Socket } from 'socket.io-client';
import type { Message } from '@/types';
import { WS_URL } from '@/lib/api-config';

let socket: Socket | null = null;

export const socketService = {
  connect: (token: string): Socket => {
    if (socket?.connected) return socket;

    socket = io(`${WS_URL}/chat`, {
      auth: { token },
      transports: ['websocket'],
    });

    return socket;
  },

  disconnect: () => {
    socket?.disconnect();
    socket = null;
  },

  joinTicket: (ticketId: string) => {
    socket?.emit('join_ticket', { ticketId });
  },

  leaveTicket: (ticketId: string) => {
    socket?.emit('leave_ticket', { ticketId });
  },

  sendMessage: (ticketId: string, content: string) => {
    socket?.emit('send_message', { ticketId, content });
  },

  onNewMessage: (callback: (message: Message) => void) => {
    socket?.on('new_message', callback);
  },

  offNewMessage: (callback: (message: Message) => void) => {
    socket?.off('new_message', callback);
  },
};

File: mobile/supporthub_app/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as io;
import '../core/config/app_config.dart';
import '../models/message.dart';

class SocketService {
  io.Socket? _socket;

  void connect(String token) {
    if (_socket?.connected == true) return;

    _socket = io.io(
      '${AppConfig.wsBaseUrl}/chat',
      io.OptionBuilder()
          .setTransports(['websocket'])
          .setAuth({'token': token})
          .enableAutoConnect()
          .build(),
    );
  }

  void disconnect() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
  }

  void joinTicket(String ticketId) {
    _socket?.emit('join_ticket', {'ticketId': ticketId});
  }

  void leaveTicket(String ticketId) {
    _socket?.emit('leave_ticket', {'ticketId': ticketId});
  }

  void sendMessage(String ticketId, String content) {
    _socket?.emit('send_message', {
      'ticketId': ticketId,
      'content': content,
    });
  }

  void onNewMessage(void Function(Message message) callback) {
    _socket?.on('new_message', (data) {
      if (data is Map<String, dynamic>) {
        callback(Message.fromJson(data));
      } else if (data is Map) {
        callback(Message.fromJson(Map<String, dynamic>.from(data)));
      }
    });
  }

  void offNewMessage() {
    _socket?.off('new_message');
  }
}
